### ISSUE TITLE
Mem Leak?

### ISSUE THREAD

**Issue by rjchicago** [traefik/plugin-log4shell#4]
Is there a possible memory leak with this plugin, or with the overall Traefik plugin system?

We're running Traefik 2.5.1 with the log4shell plugin (0.1.2) on 8 different Swarms since Dec 20 (~2 weeks). In one Swarm, Traefik is now using 30GB memory each (5 replicas). In another Swarm, Traefik is using 2GB memory each (5 replicas). The remaining 6 Swarms are seeing relatively normal memory usage - although, these past 2 weeks would also have seen much lower request volume.

### Memory
~60 nodes, 8 Swarms, memory usage over 4 weeks:
![image](https://user-images.githubusercontent.com/382450/147986939-84736246-35ca-4051-96a5-adc952363c8e.png)

I don't see anything glaring in the logs.

#### `middlewares.yml`

``` yml
http:
  middlewares:
    compress:
      compress: {}
    plugin-log4shell:
      plugin:
        log4shell:
          errorCode: "200"
```

#### `traefik.yml`

``` yml
# Traefik Config 2.x

pilot:
  token: "***********************************"

experimental:
  plugins:
    log4shell:
      moduleName: "github.com/traefik/plugin-log4shell"
      version: "v0.1.2"

# Entrypoints configuration
entryPoints:
  metrics:
    address: ':8084'
  http:
    address: ':80'
    forwardedHeaders:
      insecure: true
    http:
      middlewares:
        - compress@file
        - plugin-log4shell@file
...
```

**Comment by rjchicago** [comment]
After restarting the Traefik service in this Swarm yesterday, you can see the memory usage climbing again. This is memory usage by Traefik container over the last 16h:
![image](https://user-images.githubusercontent.com/382450/148077204-03bdb2e9-f060-4151-b18c-74e2d8580883.png)

**Comment by rjchicago** [comment]
The following chart shows Traefik container memory usage before and after removing the `log4shell` plugin configuration in this Swarm:
![image](https://user-images.githubusercontent.com/382450/148095519-e1ee6b4f-b0f5-405b-a128-9baf936cba6b.png)

**Comment by ldez** [comment]
Hello, could you try the latest release of Traefik (v2.5.6)?

**Comment by rjchicago** [comment]
Sure, testing with `v2.5.6` now. I'll leave running for about 30m and report back.

**Comment by rjchicago** [comment]
No luck with `v2.5.6` - here's with / without plugin...
![image](https://user-images.githubusercontent.com/382450/148110118-d9d4c364-3d34-43dc-960a-ce1c64c969af.png)

Also, if it helps, this is our most populous Swarm with 183 services (548 routers)...
![image](https://user-images.githubusercontent.com/382450/148110263-2f2aa24a-98cf-418e-8162-52b71db2d519.png)

**Comment by nilskohrs** [comment]
We have also had OOM problems since adding this plugin to Traefik EE 2.5.0, for now we'll remove the plugin again from our environments

**Comment by SantoDE** [comment]
@nilskohrs as I see you are on the support portal, please feel free to reach out there as well. We are currently working on the issue already.

**Comment by juliens** [comment]
Hello,

I reproduced a memory leak, but I'm not sure this is the same as yours, mine seems more link to dynamic configuration refresh.

I will need more information to go deeper in the problem.
Can you enable debug mode and give us all the pprof files when memory is low and when memory is high.

```
curl -o profile.pprof <traefik_ip>:8080/debug/pprof/profile    # CPU (takes 30 seconds to return)
curl -o heap.pprof <traefik_ip>:8080/debug/pprof/heap    
curl -o block.pprof <traefik_ip>:8080/debug/pprof/block    
curl -o mutex.pprof <traefik_ip>:8080/debug/pprof/mutex
curl -o goroutine.pprof <traefik_ip>:8080/debug/pprof/goroutine
curl -o trace.pprof <traefik_ip>:8080/debug/pprof/trace?seconds=5    # the seconds parameter is needed
```

May be you can share more information about your environment:
 - debug logs
 - how often your dynamic configuration is refresh
 - reqs/s
 - everything you think that can help

Thanks

**Comment by rjchicago** [comment]
This is our dev Swarm, so it absolutely sees the highest amount of churn. Although, it seems the memory in this swarm increases regardless of new services or containers.

We're also running many job services there that spin up one-off containers as jobs - these, however, are not enabled in traefik - though some do not have the `traefik.enable=false` label set.

I will try to get the logs for this, but this could take a couple of days to setup and coordinate as this Swarm is heavily relied upon by many teams.

Thanks

**Comment by ldez** [comment]
We found the problem and a fix, it will be available in the next release (v2.5.7) of Traefik.

**Comment by rjchicago** [comment]
@ldez, that's great news! We'll look forward to `v2.5.7` being released! Thanks for the quick response!

**Comment by rjchicago** [comment]
@ldez, just circling back to inform you that the memory issue was successfully addressed for us with `v2.5.7`. Thanks again! 👍

**Comment by ldez** [comment]
@rjchicago Thank you for your confirmation :+1:
